### ISSUE TITLE
Grpc.Core will be deprecated soon

### ISSUE THREAD

**Issue by arthurkehrwald** [VisualPinball/VisualPinball.Engine.Mpf#30]
https://grpc.io/blog/grpc-csharp-future/

We should move to [gprc-dotnet](https://github.com/grpc/grpc-dotnet)

**Comment by arthurkehrwald** [comment]
I tried to replace Grpc.Core with grpc-dotnet and couldn't get it to work. Unity lacks support for HTTP/2, which the old Grpc.Core provided through native code, but grpc-dotnet relies on .NET to do it and the version of .NET Unity runs is too old. So it seems we're stuck with Grpc.Core until Unity moves to at least .NET 6
Useful links: 
- https://learn.microsoft.com/en-us/aspnet/core/grpc/netstandard?view=aspnetcore-8.0
- https://github.com/grpc/grpc-dotnet/issues/1309
- https://github.com/doctorseus/grpc-dotnet-unity

**Comment by freezy** [comment]
At some point (planned for next major), Unity is going to move from Mono to the current .NET runtime. Maybe that will be enough?

Otherwise there seem to be a few workarounds in that other GitHub thread.

**Comment by ghost** [comment]
You can use [YetAnotherHttpHandler](https://github.com/Cysharp/YetAnotherHttpHandler) in combination with grpc-dotnet. One limitation is that Unix Domain Sockets are not supported.

**Comment by arthurkehrwald** [comment]
Thanks for the tip!

**Comment by arthurkehrwald** [comment]
I got a simple hello world client going in Unity with YetAnotherHttpHandler and grpc-dotnet. Haven't tested it with MPF yet. The question is how we package YetAnotherHttpHandler. It's a UPM package, but GRPC and a bunch of other things we need are nuget packages. The current implementation compiles all the nuget stuff into a library that links to Unity. I see three possibilities:

- We move everything into the Unity project and use https://github.com/xoofx/UnityNuGet to install all the NuGet dependencies. This would require table authors who want to use MPF to add another scoped registry though.
- We keep the NuGet stuff in a linked library and pass in the HTTP handler from the Unity side. I don't event know if this compiles but why not right? The problem is that YetAnotherHttpHandler has two NuGet dependencies of its own, so we still kind of need that scoped registry.
- We compile YetAnotherHttpHandler into the library. Most straightforward, but I think we'd have to just copy the code into this repo. Not very elegant.

@freezy What do you think?

**Comment by freezy** [comment]
I need to look into how exactly, but you can proxy stuff with [Verdaccio](https://verdaccio.org/), the registry running at registry.visualpinball.org. So we could go point one but through our own registry? What do you think?

I need to spend some time on infrastructure, GitHub just blocked the assetlib repo because of too much traffic. I'll probably set up a self-hosted Gitea instance for this, and update the Verdaccio instance at the same time.

**Comment by arthurkehrwald** [comment]
Yep sounds good. I'll set it up with UnityNuGet for now and then I'll tell you what exactly you need to proxy when it's all working.

**Comment by arthurkehrwald** [comment]
Look like @jsm174 already worked on this before. I just found his misc/grpc-dotnet branch. Why was this not merged? Did it not work?

**Comment by jsm174** [comment]
Hello! I just looked through my notes and some of my discord conversations, and I couldn't find anything on where I was at. (I think it was almost 3 years ago). I do remember that I spent countless hours on this. 😄

It's possible grpc-dotnet wasn't quite working with arm64 at the time. (I do my dev on macos).  I'm assuming that it's been so long now all that stuff is probably figured out and working.

**Comment by arthurkehrwald** [comment]
Fixed in #34
